presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java

```diff
@@ -459,9 +505,182 @@ else if (!result.getBlocked().isDone()) {
                 // get all sections ready for execution
                 .filter(this::isReadyForExecution)
                 .limit(maxConcurrentMaterializations - runningPlanSections)
+                .map(this::tryCostBasedOptimize)
                 .collect(toImmutableList());
     }
 
+    /**
+     * Utility function to invoke runtime cost-based optimizer, which determines if the probe and build side of a JoinNode should be swapped,
+     * based on the statistics of the temporary table holding materialized exchange outputs from finished children sections.
+     */
+    private StreamingPlanSection tryCostBasedOptimize(StreamingPlanSection section)
+    {
+        // no need to do runtime optimization if no materialized exchange data is utilized by the section.
+        if (!isRuntimeOptimizerEnabled(session) || getExchangeMaterializationStrategy(session) != ALL || section.getChildren().isEmpty()) {
+            return section;
+        }
+
+        // Apply runtime optimization on each StreamingSubPlan's fragment
+        Map<PlanFragment, PlanFragment> oldToNewFragment = new HashMap<>();
+        stream(forTree(StreamingSubPlan::getChildren).depthFirstPreOrder(section.getPlan()))
+                .map(StreamingSubPlan::getFragment)
+                .forEach(fragment -> {
```

Comment:
Two questions:

1. Is the ordering of depth-first required? -- Or DFS is just to make sure all the plan fragment in this section get traversed?
2. Since this lambda is a bit long, I would suggest to refactor into a static private method, e.g. `performRuntimeOptimization`

Reply:
1. The ordering is not required, it is just to make all fragments in a section traversed.

2. Sounds good!